### DISCUSSION TITLE
Applying the tranform returned by return_transform=True

### DISCUSSION THREAD

**Discussion by baruch1192** [kornia/kornia#1406] [Questions] [answered]
Hi, 
after using this code here:

```python
import torch
import kornia.augmentation as K
input = torch.rand(1, 1, 32, 32)
test = torch.rand(1, 1, 32, 32)
layer = K.RandomAffine(degrees=0, translate=(1 / 8, 1 / 8), p=0.9, return_transform=True)
output, transform = layer(input)
```

I end up with the transformation that was applied on "input" inside "transform".

The question is how do I apply the same transformation again on "test"? (without using same_on_batch=True) 
And following does it the same way with all kinds of augmentations? a.k.a:
```bash
RandomHorizontalFlip
RandomVerticalFlip
RandomDepthicalFlip
RandomRotation
RandomAffine
RandomPerspective
RandomErasing
CenterCrop
RandomCrop
RandomResizedCrop
RandomMotionBlur
```
thanks

**Comment by edgarriba** [comment]
@baruch1192 you have here an example for this use case

```python
import kornia.augmentation as K

class MyAugmentationPipeline(nn.Module):
   def __init__(self) -> None:
      super(MyAugmentationPipeline, self).__init__()
      self.aff = K.RandomAffine(
         360, return_transform=True, same_on_batch=True
      )
      self.jit = K.ColorJitter(0.2, 0.3, 0.2, 0.3, same_on_batch=True)

   def forward(self, input):
      input, transform = self.aff(input)
      input, transform = self.jit((input, transform))
      return input, transform
```

**https://kornia.readthedocs.io/en/latest/applications/image_augmentations.html#customization

**Reply by baruch1192** [reply]
@edgarriba thank you but it is not exactly what I asked.
in your case lets say I'm doing:

input = torch.rand(1, 1, 32, 32)
test = torch.rand(1, 1, 32, 32)
input_augment, transform = MyAugmentationPipeline(input)

and now I want to apply the exact same transform on test (the one I hold in transform)

how do I do that?

**Reply by edgarriba** [reply]
what's test in this context ?

**Reply by edgarriba** [reply]
what do you want to achieve ?

**Reply by baruch1192** [reply]
I just simplified the action I want to do, but I try to generate images based on real images. 
During the process, I have reconstructed images (which are real images after encode+decode).
For every reconstructed image, I have its original image in the real images.

Now I make augmentation to both real and reconstructed images > encode them and compare between them.
I want to apply the exact same augmentation on the real images as on the reconstructed ones but I don't want to enter them together as one batch. I figured the best way is to receive the transform with return_transform=True and apply it to the other group. 
The only problem is that i don't know how to apply it again.

Hope it is clearer

**Comment by edgarriba** [comment]
as described in the the link I shared, you can do:
```python
img1 = torch.rand(1,3,224,224)
img2 = torch.rand(1,3,224,224)

aug = K.RandomAffine(360, return_transform=True, same_on_batch=True)
img3, trans = aug(img1)
img4 = aug(img2, trans)
```

**Reply by baruch1192** [reply]
I'm sorry but it doesn't work. look at this example:

import torch
import kornia.augmentation as K
input1 = torch.tensor([[[[1.0, 2.0], [3.0, 4.0]]]])
input2 = torch.tensor([[[[1.0, 2.0], [3.0, 4.0]]]])
layer = K.RandomAffine(degrees=0, translate=(1 / 8, 1 / 8), p=0.9, return_transform=True)
output1, transform1 = layer(input1)
output2, transform2 = layer((input2, transform1)) # as in kornia 0.6.1 layer(input2, transform1) doesnt work

>>> print(output1)
tensor([[[[0.5801, 0.8746],
          [2.1578, 2.5939]]]])

>>> print(output2)
tensor([[[[0.3461, 0.8553],
          [1.3721, 2.6981]]]])

>>> print(transform1)
tensor([[[ 1.0000,  0.0000, -0.0702],
         [ 0.0000,  1.0000,  0.2456],
         [ 0.0000,  0.0000,  1.0000]]])

>>> print(transform2)
tensor([[[1.0000, 0.0000, 0.0899],
         [0.0000, 1.0000, 0.4911],
         [0.0000, 0.0000, 1.0000]]])


from here I'm learning that "input1" and "input2" didn't go through the same transformation. 
I want to get the result output1==output2 on this case
what am I missing?

thank you for your patience

**Comment by shijianjian** [comment, marked as answer]
There are two recommended ways of applying the exact transformation for augmentations. The direct apply a transformation matrix can be used in ```RandomAffine(...).apply_transform(image, None, trans_mat)``` functions, but I would not recommended as a standard path.

1. Using ```AugmentationSequential``` to define multiple ```input``` in ```data_keys```.
    ```python
    aug_seq = AugmentationSequential(
        RandomColorJitter(...),
        data_keys=['input', 'input']
    )
    aug_img1, aug_img2 = aug_seq(img1, img2)
    ```

2. Accessing ```_params``` in augmentations.
    ```python
    aug = RandomColorJitter(...)
    aug_img1 = aug(img1)
    aug_img2 = aug(img2, aug._params)
    ```

**Reply by baruch1192** [reply]
Thanks, it worked perfectly!!

my implementation on previous example:
import torch
import kornia.augmentation as K
input1 = torch.tensor([[[[1.0, 2.0], [3.0, 4.0]]]])
input2 = torch.tensor([[[[1.0, 2.0], [3.0, 4.0]]]])
layer = K.RandomAffine(degrees=0, translate=(1 / 8, 1 / 8), p=0.9)
output1 = layer(input1)
params = layer._params # so I can apply layer on other inputs but save the parameters for later use
output2 = layer(input2, params)

>>> print(output1)
tensor([[[[0.8002, 1.4266],
          [2.5709, 3.3117]]]])
>>> print(output2)
tensor([[[[0.8002, 1.4266],
          [2.5709, 3.3117]]]])
>>> print(output1 == output2)
tensor([[[[True, True],
          [True, True]]]])

**Comment by priyamdey** [comment]
Hi. The suggested answer 1 (using ```AugmentationSequential``` -- useful for me as I use multiple augs): I want to use the same sequential (i.e., deterministic) on multiple batches not at the same time, but in sequential manner (Eg: Batch1 comes in, apply  ```AugmentationSequential```. Batch2 comes next, apply the same params as done to Batch1). Is this possible? Thanks

**Reply by shijianjian** [reply]
It is the same to access ```_param``` attribute like ```aug_img1, aug_img2 = aug_seq(img1, img2, params=aug_seq._params)```.

**Reply by priyamdey** [reply]
Oh no sorry for not making it clear. What I meant was something like this:

```
aug_batch1 = aug_seq(batch1) # I don't have access to batch2 right now
# Say batch2 comes in
aug_batch2 = aug_seq(batch2, aug_seq._params) # This doesn't work
```

Apart from looping through augmentations in ```AugmentationSequential```, I was wondering if there is a direct way to access the same params of the sequential for the next batch. I hope it was a bit more clear now.

**Reply by shijianjian** [reply]
It should be a named parameter. Can you try:

```
aug_batch2 = aug_seq(batch2, params=aug_seq._params)
```

**Reply by priyamdey** [reply]
Oh I missed that it should be a keyword argument. Yeah it works. Thanks for clarifying!
